### ISSUE TITLE
Add Multi-Token Prediction (MTP) speculative decoding for local chat

### ISSUE THREAD

**Issue by frankhildebrandt** [frankhildebrandt/local-ai-router#28]
## What to build

Add **Multi-Token Prediction (MTP)** as a third speculative-decoding mode for local chat models, next to the existing draft-model and GGUF n-gram options.

MTP is self-speculation: the checkpoint ships a small extra prediction head that drafts 1–3 future tokens, then the main model verifies them in one forward pass. Output quality stays the same as ordinary decoding (rejection sampling on the main logits). There is no second library model to install, and RAM overhead is typically well under 10% of the target — unlike draft-model speculation, which loads a full extra sidecar model.

Typical speedups on models that actually keep the MTP tensors: about **1.5–2×** tok/s on GGUF (Qwen 3.5/3.6, `n_max` 2–3, ~75–90% second-token accept) and about **1.2–1.4×** on Apple Silicon MLX for 4B+. Tiny models (0.8B / 2B) often regress; they should not get MTP by default.

## Why this fits here

The gateway already owns speculative config, sidecar argv, RAM accounting, and the Local library editor:

- `SpeculativeMode` is only `draft_model` | `ngram` (`src-tauri/src/speculative.rs`, `src/types.ts`)
- GGUF already passes `--spec-type draft-simple` / `ngram-simple` plus `--spec-draft-n-max`
- MLX already passes `--draft-model` / `--draft-tokens` into `sidecars/mlx-server`
- The Local resource editor is the UI surface

MTP is the same product surface with a cheaper, higher-acceptance drafter **when the weights include the head**.

### GGUF / llama.cpp — ready on the current pin

`scripts/build-sidecars.sh` pins llama.cpp to `9f0d017efb4a388bd5c60a27a575c90f20868e51` (2026-08-17). That build already has MTP (`ggml-org/llama.cpp#22673`, merged 2026-05-16). Enable with:

```
--spec-type draft-mtp
--spec-draft-n-max 2
```

No `-md` draft file. Community default is `n_max=2` (sometimes 3). Prompt-processing can be a bit slower because of extra embedding transfers; decode is the win. Vision and parallel slots are supported upstream, not fully optimized.

**Hard requirement:** the GGUF must retain MTP/NextN tensors. Ordinary quants often strip them. Use explicitly MTP-labelled files, e.g. `unsloth/*-MTP-GGUF` or `ggml-org/*-MTP-GGUF`. Enabling `draft-mtp` on a file without heads must fail with a clear error, not silent next-token decode.

Linux/Windows stay CPU llama.cpp for now (#11 CUDA, #12 Vulkan). MTP still works on CPU; the interesting speedup is Metal today and CUDA/Vulkan later.

### MLX — Gemma 4 heads exist on the pin; Qwen heads need a bump

`sidecars/mlx-server` pins **mlx-swift-lm 3.31.4** (2026-06-30):

- Classic draft-model speculation: already wired via `ChatSession` + `SpeculativeDecodingConfig`
- Gemma 4 MTP: in 3.31.4 (`ml-explore/mlx-swift-lm#308`) as a **separate** `gemma4_assistant` drafter (`generate(..., mtpDrafter:)` / `MTPSpeculativeTokenIterator`). Only E4B has a published `-assistant` checkpoint; E2B / 12B do not. Catalog already lists Gemma 4 E2B / E4B / 12B.
- Qwen 3.5 / 3.6 **in-checkpoint** MTP: merged 2026-08-13 (`ml-explore/mlx-swift-lm#351`) — **after** 3.31.4. There is no newer tagged release yet. Python `mlx-lm` has `--mtp` (`ml-explore/mlx-lm#990`).

The sidecar currently turns off prefix KV cache whenever a draft container is loaded. MTP must decide the same tradeoff (prefix cache vs speculation) and document it. Vision paths already skip classic speculation.

Catalog models that become interesting once Qwen MTP lands: Qwen 3.5 4B / 9B and Qwen 3.8 27B (`qwen3_5` / `qwen3_5_moe`). Skip recommending MTP on 0.8B / 2B; measured regressions exist.

## Suggested design

Treat MTP as a first-class `SpeculativeMode::Mtp` (`"mtp"`), not a hidden default on Qwen files.

| Mode | Extra model | Engines | Default `n_max` |
| --- | --- | --- | --- |
| Draft model | yes, from library | GGUF + MLX | 16 / 5 |
| n-gram | no | GGUF only | 64 |
| **MTP** | no (heads in the same file) | GGUF now; MLX after runtime bump | **2** |

1. **Config / validation** (`speculative.rs`, `commands.rs`, `storage.rs`)
   - `mtp` needs no `draft_target_id`
   - GGUF-only until the MLX sidecar exposes `--mtp`
   - `n_max` 1–8 (or keep 1–128 but default 2; values ≥4 usually hurt)
   - RAM fingerprint: do **not** add a second `estimated_bytes`; optional ~10% MTP-layer overhead is enough
   - Mutually exclusive with draft-model. Do not stack MTP + n-gram in v1 even though llama.cpp can.

2. **GGUF sidecar argv**
   - `--spec-type draft-mtp --spec-draft-n-max <n>`
   - Startup must surface llama-server’s “no MTP tensors” error to the UI

3. **Capability detection**
   - Prefer inspect-on-install: GGUF metadata / tensor names, MLX `config.json` (`num_nextn_predict_layers` / MTP weight keys)
   - Catalog flag for curated MTP files so the editor can show the mode only when it can work
   - If detection is too expensive in v1, show MTP for all GGUF chat models and fail load with a specific message

4. **UI** (`LocalPage.tsx`)
   - Speculative mode: Off / Draft model / n-gram (GGUF) / **MTP**
   - Hide the draft-model picker for MTP
   - Hint: “Uses prediction heads in this file. Needs an MTP GGUF (or an MLX checkpoint that kept the heads). Default 2 draft tokens.”
   - Badge can stay “Speculative”

5. **MLX sidecar** (second slice)
   - Bump mlx-swift-lm to a **reviewed tag or commit** that includes #351 (do not pin `main` casually; 3.31.4 was chosen to isolate)
   - Add `--mtp` (and reuse `--draft-tokens`)
   - Load in-checkpoint MTP for Qwen 3.5+; keep prefix-cache behaviour explicit
   - Gemma 4 assistant drafter is **out of v1** (extra download, E4B-only). Revisit as a later draft-model variant if wanted.

6. **Catalog (optional, same or follow-up PR)**
   - Curated MTP GGUFs for Qwen 3.5/3.6 (Unsloth / ggml-org)
   - Do not silently replace existing non-MTP Qwen 3.5 MLX rows

## Acceptance criteria

- [ ] Local chat models can set speculative mode **MTP** without picking a draft target
- [ ] GGUF load passes `--spec-type draft-mtp --spec-draft-n-max <n>` (default 2) and does not add a second model to RAM accounting
- [ ] Enabling MTP on a GGUF **without** MTP tensors fails with a clear error; the sidecar does not start in a silent non-MTP mode
- [ ] Draft-model and n-gram keep working; MTP cannot be combined with a draft target
- [ ] A known-good MTP GGUF (Qwen 3.5 or 3.6) on macOS Metal shows higher decode tok/s than the same file with speculation off, with unchanged sampling quality at temp=0
- [ ] Docs: README speculative blurb + editor hint that MTP needs heads in the file
- [ ] Unit tests for `gguf_speculative_args` / `validate` cover the new mode
- [ ] MLX MTP is either implemented against a reviewed mlx-swift-lm bump **or** explicitly deferred in the UI (mode hidden for `mlx` until the sidecar flag exists)

## Out of scope

- Training or converting checkpoints to add MTP heads
- EAGLE-3 / DFlash / other spec types
- Stacking MTP + n-gram
- Gemma 4 `-assistant` drafter downloads
- Changing Stealth/Balanced/Performance defaults to turn MTP on automatically
- CUDA/Vulkan GGUF (covered by #11 / #12); those backends should inherit the same flags later

## References

- DeepSeek-V3 MTP (training heads reused as speculative drafter): https://arxiv.org/abs/2412.19437
- llama.cpp MTP: https://github.com/ggml-org/llama.cpp/pull/22673 (`--spec-type draft-mtp`)
- mlx-lm native MTP: https://github.com/ml-explore/mlx-lm/pull/990
- mlx-swift-lm Gemma 4 MTP: https://github.com/ml-explore/mlx-swift-lm/pull/308 (in 3.31.4)
- mlx-swift-lm Qwen 3.5/3.6 MTP: https://github.com/ml-explore/mlx-swift-lm/pull/351 (merged 2026-08-13, not in 3.31.4)
